include/grpcpp/security/tls_credentials_options.h

```diff
@@ -0,0 +1,308 @@
+ *  of the grpc_tls_server_authorization_check_arg struct in grpc_security.h for
+ *  more details. **/
+class TlsServerAuthorizationCheckArg {
+ public:
+  /** TlsServerAuthorizationCheckArg does not take ownership of the C arg passed
+   * to the constructor. One must remember to free any memory allocated to the
+   * C arg after using the setter functions below. **/
+  TlsServerAuthorizationCheckArg(grpc_tls_server_authorization_check_arg* arg);
+  ~TlsServerAuthorizationCheckArg();
+
+  /** Getters for member fields. They return the corresponding fields of the
+   * underlying C arg.**/
+  void* cb_user_data() const;
+  int success() const;
+  grpc::string target_name() const;
+  grpc::string peer_cert() const;
+  grpc_status_code status() const;
+  grpc::string error_details() const;
+
+  /** Setters for member fields. They modify the fields of the underlying C arg.
+   * The setters for target_name, peer_cert, and error_details allocate memory
+   * when modifying c_arg_, so one must remember to free c_arg_'s original
+   * target_name, peer_cert, or error_details after using the appropriate setter
+   * function.
+   * **/
+  void set_cb_user_data(void* cb_user_data);
+  void set_success(int success);
+  void set_target_name(const grpc::string& target_name);
+  void set_peer_cert(const grpc::string& peer_cert);
+  void set_status(grpc_status_code status);
+  void set_error_details(const grpc::string& error_details);
+
+  /** Calls the C arg's callback function. **/
+  void OnServerAuthorizationCheckDoneCallback();
+
+ private:
+  grpc_tls_server_authorization_check_arg* c_arg_;
+};
+
+/** An interface that the application derives and uses to instantiate a
+ * TlsServerAuthorizationCheckConfig instance. Refer to the definition of the
+ * grpc_tls_server_authorization_check_config in grpc_tls_credentials_options.h
+ * for more details on the expectations of the member functions of the
+ * interface.
+ * **/
+struct TlsServerAuthorizationCheckInterface {
+  virtual ~TlsServerAuthorizationCheckInterface() = default;
+  /** A callback that invokes the server authorization check. **/
+  virtual int Schedule(TlsServerAuthorizationCheckArg* arg) = 0;
+  /** A callback that cancels a server authorization check request. **/
+  virtual void Cancel(TlsServerAuthorizationCheckArg* arg) {}
+};
+
+/** TLS server authorization check config, wraps
+ *  grps_tls_server_authorization_check_config. It is used for experimental
+ *  purposes for now and it is subject to change. **/
+class TlsServerAuthorizationCheckConfig {
+ public:
+  TlsServerAuthorizationCheckConfig(
+      std::shared_ptr<TlsServerAuthorizationCheckInterface>
+          server_authorization_check_interface);
+  ~TlsServerAuthorizationCheckConfig();
+
+  int Schedule(TlsServerAuthorizationCheckArg* arg) const {
+    if (server_authorization_check_interface_ == nullptr) {
+      gpr_log(GPR_ERROR, "server authorization check interface is nullptr");
+      return 1;
```

Comment:
I think it will be a little confusing to return 1 as an error without populating the field `arg` since in C-core, we consider it to be aync implementation if returning non-zero. So, I would suggest to also populate other fields of  `TlsServerAuthorizationCheckArg` such as `error_details` and `status` to clearly indicate that `Schedule` API failed. I think it should apply to `TlsCredentialReloadConfig` and its members.